libcnb/src/layer/handling.rs

```diff
@@ -0,0 +1,679 @@
+            build = true
+            launch = true
+            cache = true
+            "#,
+        )
+        .unwrap();
+
+        match super::read_layer::<GenericMetadata, _, _>(&layers_dir, &layer_name) {
+            Err(ReadLayerError::LayerContentMetadataParseError(toml_error)) => {
+                assert_eq!(toml_error.line_col(), Some((1, 18)));
+            }
+            _ => panic!("Expected ReadLayerError::LayerContentMetadataParseError!"),
+        }
+    }
+
+    #[test]
+    fn read_incompatible_metadata_layer() {
+        #[derive(Deserialize, Debug, Eq, PartialEq)]
+        struct TestLayerMetadata {
+            version: String,
+            sha: String,
+        }
+
+        let layer_name = "foo";
+        let temp_dir = tempdir().unwrap();
+        let layers_dir = temp_dir.path();
+        let layer_dir = layers_dir.join(&layer_name);
+
+        fs::create_dir_all(&layer_dir).unwrap();
+        fs::write(
+            layers_dir.join(format!("{}.toml", &layer_name)),
+            r#"
+            [types]
+            build = true
+            launch = true
+            cache = true
+
+            [metadata]
+            version = "1.0"
+            "#,
+        )
+        .unwrap();
+
+        match super::read_layer::<TestLayerMetadata, _, _>(&layers_dir, &layer_name) {
+            Err(ReadLayerError::LayerContentMetadataParseError(toml_error)) => {
+                assert_eq!(toml_error.line_col(), Some((6, 12)));
+            }
+            _ => panic!("Expected ReadLayerError::LayerContentMetadataParseError!"),
+        }
+    }
+
+    #[test]
+    fn read_layer_without_layer_directory() {
+        let layer_name = "foo";
+        let temp_dir = tempdir().unwrap();
+        let layers_dir = temp_dir.path();
+        let layer_dir = layers_dir.join(&layer_name);
+
+        fs::create_dir_all(&layer_dir).unwrap();
+
+        match super::read_layer::<GenericMetadata, _, _>(&layers_dir, &layer_name) {
+            Ok(Some(layer_data)) => {
+                assert_eq!(layer_data.content_metadata, LayerContentMetadata::default());
+            }
+            _ => panic!("Expected Ok(Some(_)!"),
+        }
+    }
+
+    #[test]
+    fn read_layer_without_layer_content_metadata() {
+        let layer_name = "foo";
+        let temp_dir = tempdir().unwrap();
+        let layers_dir = temp_dir.path();
+
+        fs::write(layers_dir.join(format!("{}.toml", &layer_name)), "").unwrap();
+
+        match super::read_layer::<GenericMetadata, _, _>(&layers_dir, &layer_name) {
+            Ok(None) => {}
+            _ => panic!("Expected Ok(None)!"),
+        }
+    }
+
+    #[test]
+    fn read_nonexistent_layer() {
+        let layer_name = "foo";
+        let temp_dir = tempdir().unwrap();
+        let layers_dir = temp_dir.path();
+
+        match super::read_layer::<GenericMetadata, _, _>(&layers_dir, &layer_name) {
+            Ok(None) => {}
+            _ => panic!("Expected Ok(None)!"),
+        }
+    }
+}
```

Comment:
A stellar job with the tests. I feel confident I could modify most of this file with this suite. My one persevering desire is the ability to exercise `handle_layer` in tests. Or to otherwise see how it would be done as the end user, say in the attached ruby example app.